sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp

```diff
@@ -160,16 +160,27 @@ joint_matrix_store(Group sg,
 #endif // __SYCL_DEVICE_ONLY__
 }
 
-template <typename Group, typename T1, typename T2, size_t M, size_t K,
-          size_t N, matrix_layout LayoutA, matrix_layout LayoutB,
+template <typename Group, typename T1, typename T2, typename T3, size_t M,
+          size_t K, size_t N, matrix_layout LayoutA, matrix_layout LayoutB,
           matrix_layout LayoutC>
-inline __SYCL_ALWAYS_INLINE joint_matrix<T2, M, N, LayoutC, Group>
+inline __SYCL_ALWAYS_INLINE joint_matrix<T3, M, N, LayoutC, Group>
 joint_matrix_mad(Group sg, joint_matrix<T1, M, K, LayoutA, Group> &mA,
-                 joint_matrix<T1, K, N, LayoutB, Group> &mB,
-                 joint_matrix<T2, M, N, LayoutC, Group> &mC) {
+                 joint_matrix<T2, K, N, LayoutB, Group> &mB,
+                 joint_matrix<T3, M, N, LayoutC, Group> &mC) {
 #ifdef __SYCL_DEVICE_ONLY__
-  joint_matrix<T2, M, N, LayoutC, Group> res(sg);
-  res.spvm = __spirv_JointMatrixMadINTEL(mA.spvm, mB.spvm, mC.spvm);
+  joint_matrix<T3, M, N, LayoutC, Group> res(sg);
+  if constexpr (std::is_same<T1, uint16_t>::value &&
+                std::is_same<T2, uint16_t>::value &&
+                std::is_same<T3, float>::value)
+    res.spvm = __spirv_JointMatrixMadINTEL(mA.spvm, mB.spvm, mC.spvm);
+  else if constexpr (std::is_unsigned<T1>::value && std::is_unsigned<T2>::value)
```

Comment:
nit: may be `enable_if` is easier to recognize, then `if constexpr` , but it's up to you (or other reviewers) to change that.

Reply:
Yeah, I had thought of using enable_if but we had to create four definition for joint_matrix_mad and too much redundant code. If we use "if constexpr", the FE can optimize out branches, I did an experiment and observed this thing happen.

Reply:
'if constexpr' is better than enable_if and several funcs. The only problem with 'if constexpr' is that it is c++17 feature. Situation with c++17 is odd - we agreed that c++17 is minimum, but still try keeping SYCL header compatible with c++14.
@romanovvlad - am I correct assuming that using C++17 is Ok for new functionality/features?

Reply:
Can you share a link to this guideline? I thought we should just include an #if __cplusplus >= 201703L macro.

Reply:
>> @romanovvlad - am I correct assuming that using C++17 is Ok for new functionality/features?

It's OK if compilation of an app, which doesn't not use this feature, in C++14 mode successes.

Reply:
Since the feature is not a core one and `if constexpr` used in the templated code which is not instantiated until used I believe the patch is OK as is. We will guard the whole file if it fails during compilation in C++14 mode later.